src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs

```diff
@@ -248,17 +248,21 @@ internal void SetItem(string propertyName, JsonNode? value)
 
             OrderedDictionary<string, JsonNode?> dict = Dictionary;
 
-            if (dict.TryGetValue(propertyName, out JsonNode? replacedValue))
+            if (!dict.TryAdd(propertyName, value))
             {
+                int index = dict.IndexOf(propertyName);
```

Comment:
The duplicate lookup for the case of overwrites could be avoided assuming we had a method similar to `CollectionsMarshal.GetValueRefOrAddDefault`. Alternatively, we could consider adding an overload for `TryAdd` that returns the found index as an `out` parameter. Any particular preferences @stephentoub?

Reply:
I'd generally prefer a solution that was considered "safe" and that we could put directly onto the type.

Reply:
API proposal: https://github.com/dotnet/runtime/issues/107947